package NYT.NOrm.NServer.NObjects.NProto;

import "yt_proto/yt/core/yson/proto/protobuf_interop.proto";

import "google/protobuf/timestamp.proto";

////////////////////////////////////////////////////////////////////////////////

message TWatchRecord
{
    message TEvent
    {
        // Event type.
        optional string type = 1;

        // Serialized object key. TODO(deep): remove this after one release cycle.
        optional string object_id = 2;
        // Yson string containing object identity meta.
        optional bytes meta_yson = 5 [(NYT.NYson.NProto.yson_string) = true];

        // Serialized object labels at the moment of the event generation.
        // Used for filtering at the query side.
        optional bytes labels_yson = 3;

        message TChangedAttributes
        {
            reserved 1, 2;

            // Each character of the summary corresponds to one attribute.
            // Values: 'u' (unknown), 't' (true), 'f' (false).
            // True means the attribute was definitely affected.
            // False means it was definitely not affected.
            // Unknown means there is no information either way.
            optional string changed_summary = 3;
            optional string paths_md5 = 4;
        }

        optional TChangedAttributes changed_attributes = 4;

        message TChangedAttributeTags
        {
            message TTagsMaskPart
            {
                optional uint64 index = 1;
                optional uint64 value = 2;
            }
            repeated TTagsMaskPart tags_mask_parts = 1;
        }

        // Tags for each changed attribute.
        repeated TChangedAttributeTags changed_attribute_tags = 6;
    }

    message TTransactionContext
    {
        map<string, string> items = 1 [(NYT.NYson.NProto.yson_map) = true];
    }

    repeated TEvent events = 1;
    optional TTransactionContext transaction_context = 2;

    // Dummy events are periodically generated by masters, skipped by watch log reader.
    // They allow to seek for a given timestamp even in the absence of user-generated events.
    optional bool dummy = 3;

    optional google.protobuf.Timestamp history_time = 4;
    optional uint64 history_timestamp = 5;
}

////////////////////////////////////////////////////////////////////////////////
